server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1142,6 +1142,7 @@ public void perform(Request request, ActionListener<PrimaryResult<ReplicaRequest
                 listener = listener.map(result -> {
                     assert result.replicaRequest() == null || result.finalFailure == null
                         : "a replica request [" + result.replicaRequest() + "] with a primary failure [" + result.finalFailure + "]";
+                    assert result.replicaRequest().getParentTask().equals(request.getParentTask());
```

Comment:
I tried to address this yesterday, ran some tests, CCR failed on me and I chickened out! :-))
https://github.com/elastic/elasticsearch/pull/93042/files

I was wondering whether we we could add more to this assertion? At least a message. 

Do we need to first check that the resplica request is not null? Should we explicitly check that the parent task ID is set? I had the following in my test PR:
```
assert result.replicaRequest() == null
                        || (result.replicaRequest().getParentTask().isSet()
                            && result.replicaRequest().getParentTask().equals(request.getParentTask()))
                        : "a replica request must have the same parent task as the primary request";
```

Reply:
Ah yes that's a good point. Although this did not fail in CI - I wonder why we don't ever get a `null` here. I did as you suggest, and also strengthened the other assertion to `(result.replicaRequest() == null) != (result.finalFailure == null)` to see if that will fly.

Reply:
If it doesn't go though it means there is a **Replication**Action that in the no-failure case doesn't replicate anything!? If so, then that would nicely find cases where no `TransportReplicationAction` is needed anymore.